colexec: truncating the data and zeroing maxSetIndex is no longer sufficient for flat bytes

SHA: https://github.com/cockroachdb/cockroach/commits/0f473848083559c8a98be032949df9428068c223

Parameters:
```
TAGS=
GOFLAGS=-parallel=4
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestHashJoinerAgainstProcessor PKG=github.com/cockroachdb/cockroach/pkg/sql/distsql TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1566282&tab=buildLog

```
=== RUN   TestHashJoinerAgainstProcessor
--- join type = FULL_OUTER onExpr = "" filter = "" seed = 8984246696436413971 run = 2 ---
--- lEqCols = [1] rEqCols = [1] ---
--- inputTypes = [{{StringFamily 0 0 [] 0x667a3d0 0 <nil> [] [] 1043 <nil>}} {{StringFamily 0 0 [] 0x667a3d0 0 <nil> [] [] 1042 <nil>}}] ---
--- FAIL: TestHashJoinerAgainstProcessor (0.56s)
    columnar_operators_test.go:267: different results: no match found for row 12 of processor output
        processor output:
        		[[NULL NULL NULL NULL] [NULL NULL e'>\x05f\x04pQ' NULL] [e'h\x7fV\x0eXgQ\x0b"' '' '' ''] [NULL NULL e'\x166#' e'b\x03'] [NULL NULL 'l' e'\t4t\x13"\x01y\rv'] [NULL NULL e'\\&=' e'X%?\x0b,\x171\x05'] [NULL NULL e'B5%n\x07\'\x0f\x1cy' e'rr%$a\''] [NULL NULL 'p?{78(q' '6KPv$'] [NULL NULL e'\x16\x03Hk)<' e'>t\t_\x15'] [NULL NULL e'I\x06dC' 'Pn'] [e'h\x7fV\x0eXgQ\x0b"' '' '3 8/mluN' ''] [NULL e'\x06|\x06^v' NULL NULL] [e'W\x1aY\'U\x12d' '^' NULL NULL] [e'x<\x7f\x11' ':V' NULL NULL] [e'\t+#ZG' 'YY' NULL NULL] [e'\n\x1d/r(' e'\\dD\x01:,ad' NULL NULL] [e'&xPn\x03+D!K' e'%\x01' NULL NULL] [NULL e'>\x1688R~i' NULL NULL] [e'\x1f\x10' e'\n|)g\x1a\x1c+\x0bC' NULL NULL]]
        columnar operator output:
        		[[e'h\x7fV\x0eXgQ\x0b"' '' '' ''] [NULL NULL e'\x166#' e'b\x03'] [NULL NULL 'l' e'\t4t\x13"\x01y\rv'] [NULL NULL e'\\&=' e'X%?\x0b,\x171\x05'] [NULL NULL e'B5%n\x07\'\x0f\x1cy' e'rr%$a\''] [NULL NULL e'>\x05f\x04pQ' NULL] [NULL NULL 'p?{78(q' '6KPv$'] [NULL NULL e'\x16\x03Hk)<' e'>t\t_\x15'] [NULL NULL e'I\x06dC' 'Pn'] [e'h\x7fV\x0eXgQ\x0b"' '' '3 8/mluN' ''] [NULL e'\x06|\x06^v' NULL NULL] [NULL NULL NULL NULL] ['' '^' NULL NULL] [e'\x7f\x11' ':V' NULL NULL] [e'\t+#ZG' 'YY' NULL NULL] [e'\n\x1d/r(' e'\\dD\x01:,ad' NULL NULL] [e'&xPn\x03+D!K' e'%\x01' NULL NULL] [NULL e'>\x1688R~i' NULL NULL] [e'\x1f\x10' e'\n|)g\x1a\x1c+\x0bC' NULL NULL]]

```

This failure can be reliably reproduced with setting `seed := 5508698795671500314` in `TestHashJoinerAgainstProcessor`:
```
--- join type = LEFT_OUTER onExpr = "" filter = "" seed = 5508698795671500314 run = 1 ---
--- lEqCols = [2] rEqCols = [2] ---
--- inputTypes = [{{StringFamily 0 0 [] 0x9e19e50 0 <nil> [] [] 1042 <nil>}} {{IntFamily 16 0 [] 0x9e19e50 0 <nil> [] [] 21 <nil>}} {{StringFamily 0 0 [] 0x9e19e50 0 <nil> [] [] 25 <nil>}}] ---
--- FAIL: TestHashJoinerAgainstProcessor (0.11s)
    columnar_operators_test.go:268: different results: no match found for row 4 of processor output
        processor output:
        		[[e'\U00002603' 24437 '' '' -19605 ''] [e'\U00002603' 24437 '' 'X' -15887 ''] [NULL 1 e'2\x1b\x14^P' NULL NULL NULL] [NULL -31938 'H' NULL NULL NULL] [e'}q\x15' 733 'BO}' NULL NULL NULL] [e'<[Jg\x14' 10339 e'uV\x13' NULL NULL NULL] [e'A4\x02\x1fO:$#R' -30573 e'7\x01uRY\x18' NULL NULL NULL] [e'\'' -17161 e's\x0fF' NULL NULL NULL] [e'\x12nK' -20167 e'(\x1fT?A\'G0' NULL NULL NULL] [e'\\l\x18>(0\x02' 31380 '"' NULL NULL NULL] ['+,ep5or' 1 e'a;\x1f\t' NULL NULL NULL]]
        columnar operator output:
        		[[e'\U00002603' 24437 '' '' -19605 ''] [e'\U00002603' 24437 '' 'X' -15887 ''] [NULL 1 e'2\x1b\x14^P' NULL NULL NULL] [NULL -31938 'H' NULL NULL NULL] ['' 733 'BO}' NULL NULL NULL] [e'<[Jg\x14' 10339 e'uV\x13' NULL NULL NULL] [e'A4\x02\x1fO:$#R' -30573 e'7\x01uRY\x18' NULL NULL NULL] [e'\'' -17161 e's\x0fF' NULL NULL NULL] [e'\x12nK' -20167 e'(\x1fT?A\'G0' NULL NULL NULL] [e'\\l\x18>(0\x02' 31380 '"' NULL NULL NULL] ['+,ep5or' 1 e'a;\x1f\t' NULL NULL NULL]]
FAIL
FAIL	github.com/cockroachdb/cockroach/pkg/sql/distsql	0.172s
```
Here is what is happening during this `LEFT OUTER` join:
1. hash joiner finds two matches on the right for one of the rows from the left (the row at index 2). These two rows are output. The first column is `STRING`, and it contains the data for the both of the rows, so the contents of the flat bytes for that column are:
```
b.data = [226, 152, 131, 226, 152, 131]
b.offsets = [0, 3, 6, 0, ...]
```
2. all other 9 rows from the left do not have a match, so all the data is copied over from the build table (the left side is the build side)
2.1 first 2 values for the first column are NULLs, so the offsets are not overwritten
2.2 the remaining 7 values are copied over correctly, and we end up with the following state:
```
correct b.data
b.offsets = [0, 3, 0, 3, 8, 17, ...]
```
3 then we're setting the length on the batch which calls `UpdateOffsetsToBeNonDecreasing` which updates offsets to `b.offsets = [0, 3, 3, 3, 8, 17, ...]`, and now we're in incorrect state (we wrongly think that `b.Get(2)` is empty whereas it should be of length 3 - which is what we would get before updating the offsets).

The underlying problem is that hash joiner calls `ResetInternalBatch` which in turn calls `Reset` on all flat bytes, but that `Reset` only truncates the data and unsets `maxSetIndex`. It appears that now we have to zero out the offsets. @asubiotto thoughts?

--- a/pkg/col/coldata/bytes.go
+++ b/pkg/col/coldata/bytes.go
@@ -353,8 +353,7 @@ func (b *Bytes) Size() uintptr {
 var zeroInt32Slice = make([]int32, BatchSize())
 
 // Zero zeroes out the underlying bytes. Note that this doesn't change the
-// length. Use this instead of Reset if you need to be able to Get zeroed byte
-// slices.
+// length.
 func (b *Bytes) Zero() {
 	b.data = b.data[:0]
 	for n := 0; n < len(b.offsets); n += copy(b.offsets[n:], zeroInt32Slice) {
@@ -362,14 +361,14 @@ func (b *Bytes) Zero() {
 	b.maxSetIndex = 0
 }
 
-// Reset resets the underlying Bytes for reuse.
-// This is useful when the caller is going to overwrite the underlying bytes and
-// needs a quick way to Reset. Note that this doesn't change the length either.
+// Reset resets the underlying Bytes for reuse. Note that this zeroes out the
+// underlying bytes but doesn't change the length (see #42054 for the
+// discussion on why simply truncating b.data and setting b.maxSetIndex to 0 is
+// not sufficient).
 // TODO(asubiotto): Move towards removing Set in favor of AppendVal. At that
 //  point we can reset the length to 0.
 func (b *Bytes) Reset() {
-	b.data = b.data[:0]
-	b.maxSetIndex = 0
+	b.Zero()
 }
 
 // String is used for debugging purposes.

